Extract default column adapters to separate module

veyndan opened the pull request:
### BREAKING CHANGE

These two issues had to be fixed in conjunction, per https://github.com/cashapp/sqldelight/issues/2060#issuecomment-716022655.

#### Closes #2056.

Extracts the default column adapters for `Boolean`, `Double`, `Float`, `Int`, and `Short` into a new module named "primitive-adapters".

#### Closes #2060.

Removes the default import in the SQL file for types: `Boolean`, `Short`, `Int`, `Integer`, `Long`, `Float`, `Double`, `String`, `ByteArray`.

veyndan commented on adapters/primitive-adapters/build.gradle:
```diff
@@ -0,0 +1,12 @@
+apply plugin: 'org.jetbrains.kotlin.jvm'
+
+archivesBaseName = 'sqldelight-primitive-adapters'
```
I'm not sure how I feel about the name "primitive" tbh, as Kotlin doesn't have primitive types.

veyndan replied:
Perhaps we should name it `sqldelight-sqlite-adapters` per https://github.com/cashapp/sqldelight/pull/2121/files#r546376383?

AlecStrong replied:
yes, given our discussion yesterday the idea is that dialects will have their own repo, and part of it would be a set of adapters - and in this case these are the SQLite adapters, where something like postgres will have none of these since you should be using the correct column type :+1:

veyndan commented on adapters/primitive-adapters/gradle.properties:
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=primitive-adapters
```
I'd kinda prefer the name `adapters-primitive` as it'd look better when importing dependencies (similar to the Retrofit adapters):

```groovy
implementation "com.squareup.sqldelight:adapters-primitive:1.4.4"
implementation "com.squareup.sqldelight:adapters-postgresql:1.4.4"
```

Compared to what we have now:

```groovy
implementation "com.squareup.sqldelight:primitive-adapters:1.4.4"
implementation "com.squareup.sqldelight:postgresql-adapters:1.4.4"
```

Compared to what we have now:

AlecStrong replied:
whatever works, I think before 2.0 this would become something like "app.cash.sqldelight-sqlite:adapters"

AlecStrong replied:
or if it stays in this repo "app.cash.sqldelight:sqlite-adapters", where we're doing target first since the target is how we're grouping similar libs

veyndan commented on adapters/primitive-adapters/src/main/kotlin/com/squareup/sqldelight/adapter/primitive/BooleanColumnAdapter.kt:
```diff
@@ -0,0 +1,9 @@
+package com.squareup.sqldelight.adapter.primitive
```
So this probably should be `com.squareup.sqldelight.primitive.adapter` if we're following the approach of the other modules (e.g., `JdbcDriver` has the import `com.squareup.sqldelight.jdbc.driver`), but personally I'd find the grouping in the imports to make a lot more sense here than having the adapter name first (`com.squareup.sqldelight.adapter.primitive`).

AlecStrong replied:
similar to above, I'd say keep whatever you've already done and then we can change when we split out the dialects

veyndan commented on docs/common/custom_column_types.md:
````diff
@@ -31,6 +32,24 @@ val queryWrapper: Database = Database(
 )
 ```
 
+## Primitives
+
+A sibling module that adapts primitives for your convenience.
+
+```groovy
+dependencies {
+  implementation "com.squareup.sqldelight:primitive-adapters:{{ versions.sqldelight }}"
+}
+```
+
+The following adapters exist:
+
+- `BooleanColumnAdapter` — Retrieves `kotlin.Boolean` for an SQL type implicitly stored as `kotlin.Long`
+- `DoubleColumnAdapter` — Retrieves `kotlin.Double` for an SQL type implicitly stored as `kotlin.Long`
+- `FloatColumnAdapter` — Retrieves `kotlin.Float` for an SQL type implicitly stored as `kotlin.Long`
+- `IntColumnAdapter` — Retrieves `kotlin.Int` for an SQL type implicitly stored as `kotlin.Long`
+- `ShortColumnAdapter` — Retrieves `kotlin.Short` for an SQL type implicitly stored as `kotlin.Long`
+
````
Idk if this should be promoted for use in dialects other than SQLite, as HSQL, MySQL, and PostgreSQL all have native SQL types that map to the above (e.g., in PostgreSQL, a user shouldn't do `INT AS kotlin.Short` and use `ShortColumnAdapter` as PostgreSQL has the type `SMALLINT` which natively decodes to `kotlin.Short`). Should this be moved to the SQLite specific documentation (even though it can technically be used by other dialects)?

AlecStrong replied:
SQLite specific

veyndan commented:
Not too sure why the CI is failing. It looks like an `OutOfMemoryError`.

```
Caused by: java.lang.OutOfMemoryError: Metaspace
```

AlecStrong reviewed (approved):
would it be problematic to have these adapters live in `runtime` instead of a new module? that's what we currently do for `EnumColumnAdapter`

veyndan commented:
@AlecStrong I moved it into a separate module per your comment https://github.com/cashapp/sqldelight/issues/2056#issuecomment-713104229 (lmk if I misunderstood it).

Also, per https://github.com/cashapp/sqldelight/pull/2121#discussion_r546376383 I'm wondering whether we should promote it for usage in non-SQLite dialects as I'd imagine most other dialects shouldn't need it. Theoretically, most other dialects shouldn't need to use `EnumColumnAdapter` as there are native enum types in those dialects (but we can't support them atm due to limitations in the abstraction which can be solved in 2.0). Please let me know your thoughts regarding the other comments on the PR as they all are about this issue 😄

AlecStrong commented:
resolved comments, once its rebased I'll review all the code. And yea this should all be a separate module assuming that we go down a road where we have

```
sqlite-dialect
--adapters
--runtime
--compiler
```

and potentially we just fold adapters into runtime

veyndan commented:
@AlecStrong Took a while to fix because of bit rot and some code that was just wrong, but I've updated the PR with the following:

- Fixed adapters as they were always being encoded to `kotlin.Long` and not `kotlin.Double` in the case of `REAL`. Removed `DoubleColumnAdapter` as it doesn't really make sense.
- Moved documentation of primitive adapters to be SQLite specific.
- Made `primitive-adapters` multiplatform.
- Forbid encoding to `kotlin.Boolean` if database value isn't 0 or 1.

I left the `primitive-adapters` name per https://github.com/cashapp/sqldelight/pull/2121#discussion_r786780236.

Zlash92 commented:
Are these changes available in a release yet?

AlecStrong commented:
no, I was thinking of doing a 2.0 alpha release though - I'll do that later today

Zlash92 commented:
Ok nice. Will it be publicly available? :)

AlecStrong commented:
yep yep it'll be an official release

Zlash92 commented:
Sweet. Thanks.

Zlash92 commented:
So I can just watch out for it here: https://github.com/cashapp/sqldelight/releases ?

Zlash92 commented:
Any thought of when it will be available? 😇
